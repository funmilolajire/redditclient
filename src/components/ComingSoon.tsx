import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';

export const ComingSoon = () => {
    const boxRef = useRef<HTMLDivElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({ repeatDelay: 2 })
        tl.from([boxRef.current], {
            y: '-55vh',
            duration: 3,
            delay: 2,
            opacity: 0,
            ease: 'elastic'
        })
    })

    return (
        <div ref={boxRef} className="container flex items-center justify-center h-screen font-extrabold ComingSoon">
            <div className="bg-yellow-500 border-2 border-gray-800 border-solid skew-y-2 transform-gpu text-gray-50 text-9xl">
                <p>Coming Soon</p>
            </div>
        </div>
    )
}