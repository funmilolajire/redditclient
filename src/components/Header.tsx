import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';
import logo from '../images/logo192.png';

export const Header = () => {
    const boxRef = useRef<HTMLDivElement>(null);
    const logoRef = useRef<HTMLImageElement>(null);
    const titleRef = useRef<HTMLHeadingElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({});
        tl.from([boxRef.current], {
            duration: 1
        })
        tl.from([logoRef.current], {
            x: '-3000px',
            opacity: 0,
            ease: 'circ'
        })
        tl.from([titleRef.current], {
            x: '3000px',
            opacity: 0,
            ease: 'slow'
        })
    })

    return (
        <header>
            <div className="header-inner">
                <section ref={boxRef}>
                    <figure ref={logoRef}>
                        <img src={logo} alt="site logo" />
                    </figure>
                    <h1 ref={titleRef}>
                        <span>Reddit</span>Client
                    </h1>
                </section>
            </div>
        </header>
    )
}