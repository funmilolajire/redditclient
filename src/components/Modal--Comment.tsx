import { useRef } from 'react';
import parse from 'html-react-parser';
import { BiComment } from 'react-icons/bi';
import { BsDot } from 'react-icons/bs';
import { GoArrowDown, GoArrowUp } from 'react-icons/go';

interface ModalCommentType {
    [key: string]: any
}

export const ModalComment = ({ comment }: { comment: ModalCommentType }) => {
    const modalCommentRef = useRef<HTMLDivElement>(null);

    //parse comment html
    const parsedHTML = (html: string) => html ? parse(parse(html).toString()) : null;

    //JSX Element
    return (
        <div className="ModalComment" ref={modalCommentRef} data-permalink={comment.permalink}>
            <p className="Comment--Info">comment made <span className="time">{comment.time}</span> by <span className="author">{comment.author}</span></p>
            <div className="Comment--Body">
                {parsedHTML(comment.comment)}
            </div>
            <div className="Comment--Bottom">
                <div className="Comment--Replies">
                    <BiComment className="reply-icon icon" />
                    <span className="comments value">{comment.num_replies}</span>
                    <span className="label">{comment.num_replies !== 1 ? 'replies' : 'reply'}</span>
                </div>
                <BsDot className="dot-icon" />
                <div className="Votes Comment--Upvotes">
                    <GoArrowUp className="arrow-icon icon" />
                    <span className="upvotes value">{comment.upvotes}</span>
                    <span className="label">{comment.upvotes !== 1 ? 'upvotes' : 'upvote'}</span>
                </div>
                <BsDot className="dot-icon" />
                <div className="Votes Comment--Downvotes">
                    <GoArrowDown className="arrow-icon icon" />
                    <span className="downvotes value">{comment.downvotes}</span>
                    <span className="label">{comment.downvotes !== 1 ? 'downvotes' : 'downvote'}</span>
                </div>
            </div>
        </div>
    )
}