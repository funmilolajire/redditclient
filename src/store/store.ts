import { configureStore } from '@reduxjs/toolkit';
import subredditsReducer from './slices/subredditsSlice';
import postsReducer from './slices/postsSlice';
import moreInformationReducer from './slices/moreInformationSlice';

export const store = configureStore({
  reducer: {
    subreddits: subredditsReducer,
    posts: postsReducer,
    moreInformation: moreInformationReducer
  },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch