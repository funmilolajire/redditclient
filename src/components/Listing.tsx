import { useRef, useLayoutEffect, useCallback } from 'react';
import { useAppDispatch, useAppSelector } from '../store/hooks';
import { handleListingChange, handleTermChange } from '../store/slices/postsSlice';

export const Listing = ({ listItem }: { listItem: string }) => {
    const listingRef = useRef<HTMLLIElement>(null);

    //handle listing change and active class
    const dispatch = useAppDispatch();
    const listing = useAppSelector(state => state.posts.listing);
    const handleListing = () => {
        if (null !== listingRef.current && listing !== listingRef.current.id) {
            let newListing = listingRef.current.id;
            dispatch(handleListingChange(newListing))
            dispatch(handleTermChange(''))
        }
    }
    const handleActiveClass = useCallback(() => {
        if (null !== listingRef.current) {
            const classList = listing === listingRef.current.id ? "Listing active" : "Listing";
            listingRef.current.className = classList
        }
    }, [listing])
    useLayoutEffect(() => {
        handleActiveClass();
    }, [handleActiveClass])


    //JSX
    return (
        <li ref={listingRef} data-testid={listItem} onClick={handleListing} id={listItem}>
            {listItem}
        </li>
    )
}