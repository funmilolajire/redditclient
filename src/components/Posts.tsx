import { Post } from './Post';
import { useRef, useLayoutEffect, useEffect } from 'react';
import { getSubredditPosts, getSearchResults } from '../store/slices/postsSlice';
import { useAppDispatch, useAppSelector } from '../store/hooks';
import { gsap } from 'gsap';
import { Modal } from './Modal';

export const Posts = () => {
    // animations
    const boxRef = useRef<HTMLDivElement>(null);
    useLayoutEffect(() => {
        const tl = gsap.timeline({})
        tl.from([boxRef.current], {
            opacity: 0,
            delay: 3,
            ease: 'slow'
        })
    }, [])

    //get Posts
    const dispatch = useAppDispatch()
    let posts = useAppSelector(state => state.posts.posts)
    const listing = useAppSelector(state => state.posts.listing);
    const subreddit = useAppSelector(state => state.posts.subreddit);
    const term = useAppSelector(state => state.posts.term);
    useEffect(() => {
        window.scrollTo(0, 0);
        if (listing.length === 0 && subreddit.length === 0) {
            dispatch(getSearchResults(term))
        } else {
            if (listing.length !== 0 && subreddit.length !== 0) {
                dispatch(getSubredditPosts({ subreddit: `/${subreddit}`, listing: `/${listing}` }));
            } else if (listing.length === 0 && subreddit.length !== 0) {
                dispatch(getSubredditPosts({ subreddit: `/${subreddit}`, listing: '' }));
            } else if (listing.length !== 0 && subreddit.length === 0) {
                dispatch(getSubredditPosts({ subreddit: '', listing: `/${listing}` }));
            }
        }
    }, [subreddit, listing, term, dispatch])

    //handle modal's visibility
    const showModal = useAppSelector((state) => state.moreInformation.showModal);

    //JSX
    return (
        <>
            <div ref={boxRef} className="Posts">
                {term.length !== 0 && <p className="search--results">showing search results for: <span>{term}</span></p>}
                <>
                    {posts && posts.map((post, i) => <Post post={post} key={i} />)}
                </>
                {term.length !== 0 && posts.length === 0 && <div className="no--results">
                    <p>Sorry, it appears there are no posts matching your search.</p>
                </div>}
            </div>
            {showModal && <Modal />}
        </>
    )
}