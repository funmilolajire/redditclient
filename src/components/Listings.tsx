import { useRef, useLayoutEffect } from 'react';
import { Listing } from './Listing';
import { gsap } from 'gsap';

export const Listings = () => {
    const listings: string[] = ['hot', 'new', 'top', 'best', 'rising', 'controversial'];

    //animations
    const listingBoxRef = useRef<HTMLDivElement>(null);
    useLayoutEffect(() => {
        const tl = gsap.timeline({})
        tl.from([listingBoxRef.current], {
            x: '1200%',
            delay: 2,
            duration: 2,
            ease: 'power3.inOut'
        })
    }, [])

    return (
        <div ref={listingBoxRef} className="Listings">
            <ul className="Listings--List">
                {
                    listings && listings.map(listing => <Listing key={listing} listItem={listing} />)
                }
            </ul>
        </div>
    )
}