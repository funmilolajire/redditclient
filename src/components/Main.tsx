import { Listings } from './Listings';
import { Posts } from './Posts';
export const Main = () => (
    <div className="Main">
        <Listings />
        <Posts />
    </div>
)