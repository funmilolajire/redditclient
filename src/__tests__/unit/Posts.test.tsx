import { mount } from 'enzyme';
import { Posts } from '../../components/Posts';
import { Provider } from 'react-redux';
import { store } from '../../store/store';

window.scrollTo = jest.fn()

describe('renders Posts component', () => {
    const PostsComponent = mount(
        <Provider store={store}>
            <Posts />
        </Provider>
    )
    afterEach(() => {
        jest.resetAllMocks();
    });
    afterAll(() => {
        jest.clearAllMocks();
    });

    test('make snapshot of Posts component', () => {
        expect(PostsComponent).toMatchSnapshot();
    })
})