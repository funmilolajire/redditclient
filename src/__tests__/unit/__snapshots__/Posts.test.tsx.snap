// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Posts component make snapshot of Posts component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Posts>
    <div
      className="Posts"
    />
  </Posts>
</Provider>
`;
